test/Conversion/VectorToAIEVec/test-contract.mlir

```diff
@@ -0,0 +1,116 @@
+                            %C : vector<4x8xi32>) -> vector<4x8xi32> {
+  %0 = arith.extsi %B : vector<4x8xi8> to vector<4x8xi16>
+  %1 = vector.contract {indexing_maps = [#map1, #map2, #map3],
+                        iterator_types = ["parallel", "parallel", "reduction"],
+                        kind = #vector.kind<add>} %A, %0, %C :
+                        vector<4x4xi16>, vector<4x8xi16> into vector<4x8xi32>
+  return %1 : vector<4x8xi32>
+}
+
+// CHECK-LABEL: func.func @contracti32i16i64(
+// CHECK-SAME: %[[A:[a-zA-Z0-9]+]]: vector<4x2xi32>,
+// CHECK-SAME: %[[B:[a-zA-Z0-9]+]]: vector<2x4xi16>,
+// CHECK-SAME: %[[C:[a-zA-Z0-9]+]]: vector<4x4xi64>) -> vector<4x4xi64> {
+// CHECK:        %[[R:.*]] = aievec.matmul %[[A]], %[[B]], %[[C]] :
+// CHECK-SAME:   vector<4x2xi32>, vector<2x4xi16> into vector<4x4xi64>
+// CHECK:        return %[[R]] : vector<4x4xi64>
+func.func @contracti32i16i64(%A : vector<4x2xi32>,
+                             %B : vector<2x4xi16>,
+                             %C : vector<4x4xi64>) -> vector<4x4xi64> {
+  %0 = arith.extsi %B : vector<2x4xi16> to vector<2x4xi32>
+  %1 = vector.contract {indexing_maps = [#map1, #map2, #map3],
+                        iterator_types = ["parallel", "parallel", "reduction"],
+                        kind = #vector.kind<add>} %A, %0, %C :
+                        vector<4x2xi32>, vector<2x4xi32> into vector<4x4xi64>
+  return %1 : vector<4x4xi64>
+}
+
+// CHECK-LABEL: func.func @contractf32f32f32(
+// CHECK-SAME: %[[A:[a-zA-Z0-9]+]]: vector<4x8xbf16>,
+// CHECK-SAME: %[[B:[a-zA-Z0-9]+]]: vector<8x4xbf16>,
+// CHECK-SAME: %[[C:[a-zA-Z0-9]+]]: vector<4x4xf32>) -> vector<4x4xf32> {
+// CHECK:        %[[R:.*]] = aievec.matmul %[[A]], %[[B]], %[[C]] :
+// CHECK-SAME:   vector<4x8xbf16>, vector<8x4xbf16> into vector<4x4xf32>
+// CHECK:        return %[[R]] : vector<4x4xf32>
+func.func @contractf32f32f32(%A : vector<4x8xbf16>,
+                             %B : vector<8x4xbf16>,
+                             %C : vector<4x4xf32>) -> vector<4x4xf32> {
+  %0 = arith.extf %A : vector<4x8xbf16> to vector<4x8xf32>
+  %1 = arith.extf %B : vector<8x4xbf16> to vector<8x4xf32>
+  %2 = vector.contract {indexing_maps = [#map1, #map2, #map3],
+                        iterator_types = ["parallel", "parallel", "reduction"],
+                        kind = #vector.kind<add>} %0, %1, %C :
+                        vector<4x8xf32>, vector<8x4xf32> into vector<4x4xf32>
+  return %2 : vector<4x4xf32>
+}
+
+// CHECK-LABEL: func.func @contracti32i32i32(
+// CHECK-SAME: %[[A:[a-zA-Z0-9]+]]: vector<4x4xi16>,
+// CHECK-SAME: %[[B:[a-zA-Z0-9]+]]: vector<4x8xi8>,
+// CHECK-SAME: %[[C:[a-zA-Z0-9]+]]: vector<4x8xi32>) -> vector<4x8xi32> {
+// CHECK:        %[[R:.*]] = aievec.matmul %[[A]], %[[B]], %[[C]] :
+// CHECK-SAME:   vector<4x4xi16>, vector<4x8xi8> into vector<4x8xi32>
+// CHECK:        return %[[R]] : vector<4x8xi32>
+func.func @contracti32i32i32(%A : vector<4x4xi16>,
```

Comment:
I see, the version first promotes the operands to i32 whereas the version at L31 promotes the i8 operand to i16.

Reply:
I'm dealing with two issues here. First, when you generalize `linalg.matmul` with mixed operands, it generates widening operations before the arithmetic, and when you vectorize you end up with something like what you see there. At the same time, although the `vector.contract` operation supports different precision for inputs and outputs, it doesn't support different precision between the two inputs (A & B must have the same elemental type). This is me just making sure that there's a lowering strategy for all our supported cases 🙂